### PR TITLE
metadata: fix auto-resizing of textview widgets

### LINKED ISSUE #10582: Metadata Editor: Wrap Long Fields
When using the metadata editor in the lighttable it is possible to increase the vertical size of the data entry fields. However, when entering long text, the box often instead increases _horizontal_ size until the labels are ellipsized to nothing and then continues to increase until the right-hand-side of the right-hand panel starts disappearing off the edge of the screen. 

Screenshot_2021-12-07_14-14-21

As soon as I mouse over another image the screen refreshes and the text appears correctly.

Screenshot_2021-12-07_14-15-58

### PR DESCRIPTION
**Pull request by elstoc**
Resolves #10582.

### CODE REVIEW

**Comment by phweyland** [comment]
Definitively avoids the bug.
But prevents ellipsizing the labels.
image

GTK_POLICY_EXTERNAL seems to behave the same minus the horizontal scrollbar:
image


if @dterrahe has no better idea this may be a workaround.

**Comment by dterrahe** [comment]
> But prevents ellipsizing the labels

I think this was tried and rejected before for this reason; it doesn't work too well with narrow panels.

**Comment by elstoc** [comment]
> it doesn't work too well with narrow panels.

But on the other hand if I had wide panels I wouldn't have seen the bug in the first place. In my opinion (and I love narrow panels), the issue reported in #10582 is worse. And I really can't find any other solution, apart from drawing the whole thing manually.

**Comment by dterrahe** [comment]
> the issue reported in #10582 is worse

That's debatable. That issue shows up while you're typing which is definitely very annoying if you don't touch type, but temporary. If the labels make the entries really narrow, you cannot see the metadata while moving around, so you can't spot errors without scrolling through each entry. Let's disagree on severity. _My_ preference would be to leave as is until someone finds a solution. Just one vote, obviously.

**Comment by elstoc** [comment]
The alternative is to change it to 
```
gtk_scrolled_window_set_policy(GTK_SCROLLED_WINDOW(swindow), GTK_POLICY_NEVER, GTK_POLICY_ALWAYS);
```
This leave ellipsization intact and seems to prevent the bug but means that scroll bars are always displayed in every entry widget.

**Comment by phweyland** [comment]
image

I prefer this one. The less intrusive one.

To break a word longer than the line, as above, change GTK_WRAP_WORD to GTK_WRAP_WORD_CHAR three lines below.

File: src/libs/metadata.c [hunk 1]
```diff
@@ -745,6 +745,7 @@ void gui_init(dt_lib_module_t *self)
     gtk_grid_attach(grid, swindow, 1, i, 1, 1);
     gtk_widget_set_hexpand(swindow, TRUE);
     d->swindow[i] = swindow;
+    gtk_scrolled_window_set_policy(GTK_SCROLLED_WINDOW(swindow), GTK_POLICY_AUTOMATIC, GTK_POLICY_AUTOMATIC);
     gtk_widget_set_size_request(d->swindow[i], -1, DT_PIXEL_APPLY_DPI(30));
 
     gtk_text_view_set_wrap_mode(GTK_TEXT_VIEW(textview), GTK_WRAP_WORD);
```
Comment (phweyland) [inline src/libs/metadata.c:751]: Changing GTK_WRAP_WORD to GTK_WRAP_WORD_CHAR you fix also the case of very long word.
Sorry, 3 lines below

**Review by phweyland — CHANGES_REQUESTED** [review]

**Comment by elstoc** [comment]
Latest commit implements the last two suggestions. Perhaps we should consider this a workaround for now and leave the issue open for investigation (or raise a new one).

**Review by phweyland — APPROVED** [review]
For me that's the best we can do for now.

**Comment by dterrahe** [comment]
> The alternative is to change

I know you want to do something to fix the bug and this is doing something, but all of these options (as well as the _CHAR) were considered at the time and the status quo was picked as the least bad compromise. I don't see a reason to change my view on this, but would be happy to see more feedback.

Since this is not a regression from 3.6.1 and not a crash bug, would it really be appropriate for 3.8, with very little time to collect such feedback?

**Comment by elstoc** [comment]
@dterrahe can you provide a link to the original discussion please

**Comment by phweyland** [comment]
> but all of these options (as well as the _CHAR) were considered at the time and the status quo was picked as the least bad compromise. I don't see a reason to change my view on this, but would be happy to see more feedback.

I may be wrong (my memory may fail unfortunately), but I don't remember nor having discussed these options neither having been happy with this really annoying issue. Do you have any link to those discussions ?

**Comment by dterrahe** [comment]
Don't get me wrong, I'm not claiming there was a lot of discussion of all the options (people were sort of getting fed up with "narrow panels") or that some sort of consensus was reached, so let me rephrase:

> I clearly remember experimenting with all of the fixes suggested here so far and _I_ found all of them worse than the status quo, so _I_ rejected them, not because it wasn't me who didn't come up with them, but even though I did.

Of course that doesn't mean I can't have been wrong about that. But I still feel the same way, so its 2-1 ;-)

Which, in a democracy, means this will probably go against me, which is ok! I just wanted to make the argument that, with narrow panels, _always_ wasting space on vertical scrollbars, when you can expand the height of the textboxes so that you actually _never_ need to scroll, is an unwelcome tradeoff to me. But the bug is very annoying too, so...

BTW the first time this bug was reported it was identified as a regression in GTK https://github.com/darktable-org/darktable/issues/9172 A bug report was filed https://gitlab.gnome.org/GNOME/gtk/-/issues/4042 but not followed up on. Maybe that would be the correcter way forward? Clearly the response that no changes were made _in GtkTextView_ between 3.24.24 and 3.24.29 is insufficient, in that the bug was triggered by simply upgrading gtk. There is a subtle interaction here between wrapping and scrollbar, so indeed the change could well have been elsewhere but only triggered unwanted behavior in TextViews.

**Comment by dterrahe** [comment]
Please test if #10589 also gets rid of the bug. It only adds the scrollbar when it is needed, i.e. when the textbox is actually being edited. Which takes care of my main complaint against this approach.

It also generically applies the fix to _all_ scrollwrapped textviews, not just metadata. Which is a good thing. Unless it triggers bugs elsewhere. Hence why needs testing :-)

**Comment by dterrahe** [comment]
Scratch that; metadata is the only place where textviews get scrollwrapped, so this should indeed be local to it. And it also conveniently already has focus in/out callbacks, so #10589 expands the workaround by just one line (and comments).

**Comment by phweyland** [comment]
@dterrahe thank you for the explanations. 
I understand that wasting space is not desirable, but I would vote here for this fix, as the waste is not so much and still better than showing buggy behaviors.

**Comment by dterrahe** [comment]
> I would vote here for this ~fix~ workaround

I understand your position.

What I would suggest, if this gets merged, is to have workarounds for specific gtk issues be clearly marked as such in source commentary. That way, there's at least a chance that the workaround can be removed if the cause is fixed. Otherwise, we'll end up with a lot of cruft and suboptimal choices that we can never get rid of, because we don't even know what to test to see if it is still needed. Not _as_ necessary in cases where the workaround doesn't cause any other issues, but in this case, we could easily see someone in future reverting the _ALWAYS because, after all, _AUTOMATIC is the more customary setting.

Yes, git blame will point to this PR, but only until the line gets moved/reformatted/copied all over the place by people who don't understand the reason for the magic invocations but copy them along out of an abundance of caution.

**Comment by elstoc** [comment]
Comment added

**Review by TurboGit — APPROVED** [review]
Not great to have the scrollbar always on, but better than having the GUI going crazy when entering long text in there!
